fdbserver/DataDistribution.actor.cpp

```diff
@@ -2968,6 +2967,82 @@ struct DDTeamCollection : ReferenceCounted<DDTeamCollection> {
 	}
 };
 
+// add server to wiggling queue
+void StorageWiggler::addServer(const UID& serverId, const StorageMetadataType& metadata) {
+	// std::cout << "size: " << pq_handles.size() << " add " << serverId.toString() << " DC: "<< teamCollection->primary
+	// << std::endl;
+	ASSERT(!pq_handles.count(serverId));
+	pq_handles[serverId] = wiggle_pq.emplace(metadata, serverId);
+	nonEmpty.set(true);
+}
+void StorageWiggler::removeServer(const UID& serverId) {
+	// std::cout << "size: " << pq_handles.size() << " remove " << serverId.toString() << " DC: "<<
+	// teamCollection->primary <<std::endl;
+	if (contains(serverId)) { // server haven't been popped
+		auto handle = pq_handles.at(serverId);
+		pq_handles.erase(serverId);
+		wiggle_pq.erase(handle);
+	}
+	nonEmpty.set(!wiggle_pq.empty());
+}
+void StorageWiggler::updateMetadata(const UID& serverId, const StorageMetadataType& metadata) {
+	//	std::cout << "size: " << pq_handles.size() << " update " << serverId.toString()
+	//	          << " DC: " << teamCollection->primary << std::endl;
+	auto handle = pq_handles.at(serverId);
+	if ((*handle).first.expireNow == metadata.expireNow && (*handle).first.createdTime == metadata.createdTime) {
+		return;
+	}
+	wiggle_pq.update(handle, std::make_pair(metadata, serverId));
+}
+Optional<UID> StorageWiggler::getNextServerId() {
+	if (!wiggle_pq.empty()) {
+		auto [metadata, id] = wiggle_pq.top();
+		wiggle_pq.pop();
+		pq_handles.erase(id);
+		return Optional<UID>(id);
+	}
+	return Optional<UID>();
+}
+
+Future<Void> StorageWiggler::resetStats() {
+	auto newMetrics = StorageWiggleMetrics();
+	newMetrics.smoothed_round_duration = metrics.smoothed_round_duration;
+	newMetrics.smoothed_step_duration = metrics.smoothed_step_duration;
+	metrics = newMetrics;
+	return metrics.runSetTransaction(teamCollection->cx, teamCollection->primary);
+}
+
+Future<Void> StorageWiggler::restoreStats() {
+	auto& metricsRef = metrics;
+	auto assignFunc = [&metricsRef](Optional<Value> v) {
+		if (v.present()) {
+			metricsRef = BinaryReader::fromStringRef<StorageWiggleMetrics>(v.get(), IncludeVersion());
+		}
+		return Void();
+	};
+	auto readFuture = metrics.runGetTransaction(teamCollection->cx, teamCollection->primary);
+	return map(readFuture, assignFunc);
+}
+Future<Void> StorageWiggler::startStep() {
```

Comment:
I thought during the code walkthrough meeting, we decided to avoid using "Step" and replace with "Wiggle" or something similar?